### PR TITLE
Added basic async request handling.

### LINKED ISSUE #16: Refactor result from janus
Instead of data only, result should contain ["plugindata", "jsep", "sender"]

### PR DESCRIPTION
**Pull request by Hajto**

### CODE REVIEW

**Comment by Hajto** [comment]
I went with storing type of request instead of response counter.

File: lib/connection/transaction.ex [hunk 1]
```diff
@@ -152,4 +161,8 @@ defmodule Janus.Connection.Transaction do
     }, data = #{inspect(data)}"
     |> Logger.warn()
   end
+
+  defp should_delete?(response, type)
+  defp should_delete?({:ok, %{"janus" => "ack"}}, :async_request), do: true
```
Comment (bblaszkow06) [inline lib/connection/transaction.ex]: I think we should not delete the transaction on ack of async request

Comment (bblaszkow06) [reply lib/connection/transaction.ex]: Make sure the tests cover that case

File: lib/connection.ex [hunk 2]
```diff
@@ -106,7 +106,11 @@ defmodule Janus.Connection do
   """
   @spec call(GenServer.server(), map, timeout) :: {:ok, any} | {:error, any}
   def call(server, payload, timeout \\ @default_timeout) do
-    GenServer.call(server, {:call, payload, timeout}, timeout)
+    GenServer.call(server, {:call, payload, timeout, :sync_request}, timeout)
+  end
+
+  def call_async(server, payload, timeout \\ @default_timeout) do
+    GenServer.call(server, {:call, payload, timeout, :async_request}, timeout)
```
Comment (bblaszkow06) [inline lib/connection.ex]: Remember to expose this in Session

File: lib/connection/transaction.ex [hunk 3]
```diff
@@ -118,9 +124,12 @@ defmodule Janus.Connection.Transaction do
     transaction_status = transaction_status(pending_calls_table, transaction)
 
     case transaction_status do
-      {:ok, from} ->
+      {:ok, {from, type}} ->
         GenServer.reply(from, response)
-        :ets.delete(pending_calls_table, transaction)
+
+        if should_delete?(response, type) do
+          :ets.delete(pending_calls_table, transaction)
```
Comment (bblaszkow06) [inline lib/connection/transaction.ex:136]: Actual deletion should be logged

File: test/connection/transaction_test.exs [hunk 4]
```diff
@@ -82,28 +96,62 @@ defmodule Janus.Connection.TransactionTest do
...
+    test "ignores ack", %{table: table, transaction: transaction} do
+      response = {:ok, %{"janus" => "ack"}}
+
+      logs =
+        capture_log(fn ->
+          Transaction.handle_transaction(response, transaction, table, @now)
+        end)
+
+      assert logs =~ "Call OK:"
+      # Ensure transaction is deleted
```
Comment (bblaszkow06) [inline test/connection/transaction_test.exs]: Is it deleted?

File: test/session_test.exs [hunk 5]
```diff
@@ -63,6 +88,27 @@ defmodule Janus.SessionTest do
                Session.execute_request(session, %{janus: :test})
     end
 
+    test "ignore ack", %{connection: conn} do
```
Comment (bblaszkow06) [inline test/session_test.exs:91]: We're missing test cases for two responses arriving - [ack, event] and [event, ack]

File: lib/connection/transaction.ex (hunk 3 above)
Comment (bblaszkow06) [reply lib/connection/transaction.ex:136]: We can change the way logs work here - please log from every case, this way without additional log lines we can include info about what happened (e.g. something was or wasn't removed from ets table)

File: lib/connection.ex [hunk 6]
```diff
@@ -519,6 +520,27 @@ defmodule Janus.Connection do
...
+  defp handle_payload(
+         %{
+           "janus" => "event",
+           "transaction" => transaction,
+           "plugindata" => %{
+             "plugin" => _plugin,
+             "data" => data
+           },
+           "jsep" => jsep,
+           "sender" => sender
```
Comment (bblaszkow06) [inline lib/connection.ex]: As mentioned in person, this fields will only be included in some requests, so we cannot match on them here

**Review by bblaszkow06 — CHANGES_REQUESTED** [review]

File: lib/connection.ex (hunk 6 above)
Comment (bblaszkow06) [reply lib/connection.ex]: I've decided to stick to the Janus original payload shape and only filter out transaction. Modifying the returned payload has to be done in a bigger refactor allowing to document such shape. See #17

File: test/session_test.exs (hunk 5 above)
Comment (bblaszkow06) [reply test/session_test.exs:91]: Left this as TODO since it's tested in lower layers

**Pull request merged**